### PR TITLE
DDF-2851 Added a file pattern blacklist to the content directory monitor

### PR DESCRIPTION
**Pull request by bdeining**
#### What does this PR do?
This PR reads the bad.files.* properties in the system.properties file and creates a blacklist for ingesting files through the CDM.
#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged)?
@glenhein @rzwiefel 
#### Select at least one member from relevant component team(s) from below (at least one component team member needs to approve the PR).
@Lambeaux 
#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@jlcsmith
@kcwire
#### How should this be tested? (List steps with links to updated documentation)
Build / Install / Configure CDM / Observe blacklisted file types
#### What are the relevant tickets?
[DDF-2851](https://codice.atlassian.net/browse/DDF-2851)
#### Checklist:
- [X] Documentation Updated
- [ ] Change Log Updated
- [X] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: catalog/core/catalog-core-directorymonitor/src/main/resources/OSGI-INF/metatype/metatype.xml [hunk 1]
```diff
@@ -38,10 +38,14 @@
                     label="Monitor in place" value="in_place"/>
         </AD>
 
+        <AD description="Optional: Blacklisted file pattens.  These are regular expressions used to exclude different file patterns from being ingested.  The default values are temporary Windows files (*.tmp) and Windows thumbnails (thumbs.db)."
+            name="Blacklisted File Patterns" id="blacklistedFilePatterns" required="false" type="String"
+            cardinality="100"
+            default=".*.tmp,thumbs.db"/>
```
Comment (jlcsmith) [inline catalog/core/catalog-core-directorymonitor/src/main/resources/OSGI-INF/metatype/metatype.xml]: The Camel File Component docs say that the exclude property is case sensitive.  Isn't it "Thumbs.db" not "thumbs.db"?   The documentation is lacking here.  It looks like you might be able to make it case insensitive using 
`(?i)thumbs.db`

Comment (bdeining) [inline catalog/core/catalog-core-directorymonitor/src/main/resources/OSGI-INF/metatype/metatype.xml]: According to the documentation, "matching is case in-sensitive from Camel 2.17 onward".  We currently use 2.18, so this shouldn't be an issue.  http://camel.apache.org/file2.html

Comment (jlcsmith) [inline catalog/core/catalog-core-directorymonitor/src/main/resources/OSGI-INF/metatype/metatype.xml]: 👍

File: catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitor.java [hunk 2]
```diff
@@ -268,6 +278,22 @@ public void setAttributeOverrides(List<String> attributeOverrides) {
         this.attributeOverrides = attributeOverrideMap;
     }
 
+    public void setBlacklistedFilePatterns(List<String> blacklistedFilePatterns) {
```
Comment (stustison) [inline catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitor.java]: this has security implications as well and should probably use the list of file extensions that we already have: https://github.com/codice/ddf/blob/master/distribution/ddf-common/src/main/resources/etc/system.properties#L118

You can add the thumbnail and tmp files to that list instead.

Comment (bdeining) [inline catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitor.java]: Changed the CDM to use the `bad.file.extensions` and `bad.files` properties instead

File: catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitor.java [hunk 3]
```diff
@@ -117,6 +124,17 @@ public ContentDirectoryMonitor(CamelContext camelContext, int maxRetries,
...
         this.delayBetweenRetries = delayBetweenRetries;
         this.configurationExecutor = configurationExecutor;
+        setBlacklist();
+    }
+
+    private void setBlacklist() {
+        String badFileProperty = System.getProperty("bad.files");
+        String badFileExtensionProperty = System.getProperty("bad.file.extensions");
+        badFiles = StringUtils.isNotEmpty(badFileProperty) ?
+                Arrays.asList(badFileProperty.split(",")) :
```
Comment (glenhein) [inline catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitor.java]: I'm wondering if these split regex'es should include optional whitespace before and after the comma. (\s*,\s*).

File: catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitor.java [hunk 4]
```diff
@@ -117,6 +124,17 @@ public ContentDirectoryMonitor(CamelContext camelContext, int maxRetries,
         this.maxRetries = maxRetries;
         this.delayBetweenRetries = delayBetweenRetries;
         this.configurationExecutor = configurationExecutor;
+        setBlacklist();
+    }
+
+    private void setBlacklist() {
+        String badFileProperty = System.getProperty("bad.files");
```
Comment (jlcsmith) [inline catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitor.java]: do we have a DDF constant for these props?

File: distribution/docs/src/main/resources/_contents/_resources/content-directory-monitor-contents.adoc [hunk 5]
```diff
@@ -72,6 +72,11 @@ To set multi-valued attributes, use a separate override for each value. For exam
 `*topic.keyword=PPI*` +
 `*topic.keyword=radiology*`
 
+.Blacklist
```
Comment (jlcsmith) [inline distribution/docs/src/main/resources/_contents/_resources/content-directory-monitor-contents.adoc]: This needs to be updated.  I think it would be worthwhile mention that the CDM now uses the system properties.

File: catalog/core/catalog-core-directorymonitor/src/test/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitorTest.java [hunk 6]
```diff
@@ -135,6 +155,25 @@ public void testMoveFile() throws Exception {
     }
 
     @Test
+    public void testBlackListedFileExtension() throws Exception {
+        System.setProperty("bad.file.extensions", ".txt");
```
Comment (glenhein) [inline catalog/core/catalog-core-directorymonitor/src/test/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitorTest.java]: Could you have a test that includes a comma-separated list of extensions?

Comment (bdeining) [inline catalog/core/catalog-core-directorymonitor/src/test/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitorTest.java]: Added a test to cover a comma-separated list

File: catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitor.java (hunk 3 above)
Comment (bdeining) [inline catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitor.java]: I'm not sure it's necessary, as this is property is set by reading https://github.com/codice/ddf/blob/master/distribution/ddf-common/src/main/resources/etc/system.properties#L118

File: catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitor.java (hunk 4 above)
Comment (bdeining) [inline catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitor.java]: It appears there is no constant for these properties

File: distribution/docs/src/main/resources/_contents/_resources/content-directory-monitor-contents.adoc (hunk 5 above)
Comment (bdeining) [inline distribution/docs/src/main/resources/_contents/_resources/content-directory-monitor-contents.adoc]: Updated docs

File: distribution/docs/src/main/resources/_contents/_resources/content-directory-monitor-contents.adoc [hunk 7]
```diff
@@ -72,6 +72,12 @@ To set multi-valued attributes, use a separate override for each value. For exam
...
+The CDM blacklist uses the "bad.files" and "bad.file.extensions" properties from the system.properties file in "etc/" in order to prevent
+malicious or unwanted data from being ingested into DDF.  While the CDM will automatically omit hidden files, this is particularly useful when
+an operating system automatically generated files that should not be ingested.  One such example of this is "thumbs.db" in Windows.  This file type, as well as
```
Comment (Lambeaux) [inline distribution/docs/src/main/resources/_contents/_resources/content-directory-monitor-contents.adoc]: Do not use past-tense. Re-word to: `...automatically generates files...`

File: distribution/docs/src/main/resources/_contents/_resources/content-directory-monitor-contents.adoc [hunk 8]
```diff
@@ -72,6 +72,12 @@ To set multi-valued attributes, use a separate override for each value. For exam
...
+temporary files are included in the blacklist.
```
Comment (Lambeaux) [inline distribution/docs/src/main/resources/_contents/_resources/content-directory-monitor-contents.adoc]: Oxford comma needed: `...This file type, as well as temporary files, are included in the blacklist. `

*OR*

Re-reword and get rid of commas completely (better): `This file type and any temporary files are included in the blacklist.` (as long as there does *not* exist temp files that *can* be ingested)

File: catalog/core/catalog-core-directorymonitor/src/test/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitorTest.java [hunk 9]
```diff
@@ -135,6 +155,35 @@ public void testMoveFile() throws Exception {
...
+
+    @Test
+    public void testBlackListedFileExtensions() throws Exception {
+        System.setProperty("bad.file.extensions", ".txt,.bin");
+        monitor = createContentDirectoryMonitor();
+        submitConfigOptions(monitor, monitoredDirectoryPath, ContentDirectoryMonitor.MOVE);
+        doAndVerifyFileDidNotMove(monitoredDirectory, monitoredDirectory, "input1.txt");
+        doAndVerifyFileDidNotMove(monitoredDirectory, monitoredDirectory, "input1.bin");
+        System.setProperty("bad.file.extensions", "");
+    }
```
Comment (Lambeaux) [inline catalog/core/catalog-core-directorymonitor/src/test/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitorTest.java]: ❗️ Missing a test where the properties are present but empty, and will result in state variables being null.

File: catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitor.java (hunk 4 above)
Comment (bdeining) [inline catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitor.java]: `bad.files` and `bad.file.extensions` are existing properties in the `system.properties` file for  karaf.  These are automatically made available system properties during the karaf boot process.  We've decided to use these properties because other endpoints use the same list; we want the CDM to follow the same conventions when validating files for ingest.

File: catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitor.java [hunk 10]
```diff
@@ -268,6 +288,27 @@ public void setAttributeOverrides(List<String> attributeOverrides) {
         this.attributeOverrides = attributeOverrideMap;
     }
 
+    private String getBlackListedFilePatternString() {
```
Comment (bdeining) [inline catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitor.java]: Renamed the method to `getBlackListAsRegex` to be more clear what this is doing.

File: catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitor.java [hunk 11]
```diff
@@ -268,6 +288,27 @@ public void setAttributeOverrides(List<String> attributeOverrides) {
...
+        List<String> patterns = new ArrayList<>();
+
+        if (!CollectionUtils.isEmpty(badFileExtensions)) {
```
Comment (bdeining) [inline catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitor.java]: Done

File: catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitor.java [hunk 12]
```diff
@@ -268,6 +288,27 @@ public void setAttributeOverrides(List<String> attributeOverrides) {
...
+            patterns.addAll(badFileExtensions.stream()
+                    .map(s -> (".*" + s))
+                    .collect(Collectors.toList()));
+        }
+
+        if (!CollectionUtils.isEmpty(badFiles)) {
```
Comment (bdeining) [inline catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitor.java]: Done

File: distribution/docs/src/main/resources/_contents/_resources/content-directory-monitor-contents.adoc (hunk 7 above)
Comment (bdeining) [inline distribution/docs/src/main/resources/_contents/_resources/content-directory-monitor-contents.adoc]: Done

File: distribution/docs/src/main/resources/_contents/_resources/content-directory-monitor-contents.adoc (hunk 8 above)
Comment (bdeining) [inline distribution/docs/src/main/resources/_contents/_resources/content-directory-monitor-contents.adoc]: Done

File: catalog/core/catalog-core-directorymonitor/src/test/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitorTest.java (hunk 9 above)
Comment (bdeining) [inline catalog/core/catalog-core-directorymonitor/src/test/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitorTest.java]: Done

File: catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitor.java (hunk 3 above)
Comment (bdeining) [inline catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/ContentDirectoryMonitor.java]: Done

**Comment by jlcsmith** [comment]
built and verified CDM honors bad.files and bad.file.extensions properties
